src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +48,55 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    if (totalAH_>BATTERY_AMP_HOUR_CAPACITY)
+        return 0;
+    else if (totalAH_<0)
+        return BATTERY_AMP_HOUR_CAPACITY;
+    else
+        return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime null_time(0,0,0,0);
+    if (return_time.isNull())
+        return null_time;
+    else
+        return return_time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (previousTime_.isNull())
+    {
+        totalAH_=BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+        previousTime_=batteryData.time;
+    }
+    zero_time.setHMS(0,0,0,0);
+    deltaTime_=previousTime_.msecsTo(batteryData.time)/HOUR_TO_SEC;
+    deltaCurrent_=(batteryData.current+previousCurrent_)/2;
+    amph_=deltaTime_*deltaCurrent_;
+    totalAH_+=amph_;
+
+    if (batteryData.current < 0){
+        isCharging_=true;
```

Comment:
Leaving this functionality to the designated function would be clearer. So instead of 
```
bool BatteryStateOfChargeService::isCharging() const
{
    return isCharging_;
}
```
We can remove a few lines of code without sacrificing readability with:
```
bool BatteryStateOfChargeService::isCharging() const
{
    return previousCurrent_ < 0;
}
```